functions/src/on-create-loi.ts

```diff
@@ -0,0 +1,100 @@
+/**
+ * Copyright 2024 The Ground Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {EventContext} from 'firebase-functions';
+import {QueryDocumentSnapshot} from 'firebase-functions/v1/firestore';
+import {db} from '@/common/context';
+import {Datastore} from '@/common/datastore';
+import {broadcastSurveyUpdate} from '@/common/broadcast-survey-update';
+import {geojsonToWKT} from '@terraformer/wkt';
+
+type PropertyGenerator = {
+  name: string;
+  prefix: string;
+  url: string;
+};
+
+export async function onCreateLoiHandler(
+  snapshot: QueryDocumentSnapshot,
+  context: EventContext
+) {
+  const surveyId = context.params.surveyId;
+  const loiId = context.params.loiId;
+  const loi = snapshot.data();
+
+  if (!loiId || !loi) return;
+
+  let properties = loi.properties || {};
+
+  const propertyGenerators = await db.fetchPropertyGenerators();
+
+  await Promise.all(
+    propertyGenerators.docs.map(async propertyGeneratorDoc => {
+      const propertyGenerator = propertyGeneratorDoc.data();
+
+      const {url, prefix} = propertyGenerator as PropertyGenerator;
+
+      if (prefix) properties = removePrefixedKeys(properties, prefix);
+
+      if (url) {
+        const wkt = geojsonToWKT(Datastore.fromFirestoreMap(loi.geometry));
+
+        const newProperties = await fetchProperties(url, wkt);
+
+        properties = {
+          ...properties,
+          ...(prefix ? prefixKeys(newProperties, prefix) : newProperties),
+        };
+      }
+    })
+  );
+
+  await db.updateLoiProperties(surveyId, loiId, properties);
+
+  await broadcastSurveyUpdate(context.params.surveyId);
+}
+
+type Properties = {[key: string]: string | number};
+
+async function fetchProperties(url: string, wkt: string): Promise<Properties> {
+  const response = await fetch(url, {
+    method: 'POST',
+    headers: {
+      'Content-Type': 'application/json',
+    },
+    body: JSON.stringify({wkt}),
+  });
+
+  if (!response.ok) return {};
+
+  const json = await response.json();
+
+  return json?.data[0] || {};
```

Comment:
Please add comment explaining that we expect data to be returned as the first element of an array in a field under JSON root called `data`.